pulsar-transaction/coordinator/src/main/java/org/apache/pulsar/transaction/coordinator/impl/TxnLogBufferedWriter.java

```diff
@@ -217,9 +224,9 @@ public void asyncAddData(T data, AddDataCallback callback, Object ctx){
                 // Avoid missing callback, do failed callback when error occur before add data to the array.
                 int recordsCountAfter = dataArray.size();
                 if (recordsCountAfter == recordsCountBeforeAdd){
-                    callback.addFailed(new ManagedLedgerException.ManagedLedgerFencedException(e), ctx);
+                    callback.addFailed(new ManagedLedgerInterceptException(e), ctx);
```

Comment:
Your condition above to only call the callback if number of records in data array haven't changed has a bug.
Look in internalAsyncAddData:
```java
        if (dataLength >= batchedWriteMaxSize){
            trigFlushByLargeSingleData();
            ByteBuf byteBuf = dataSerializer.serialize(data);
            managedLedger.asyncAddEntry(byteBuf, DisabledBatchCallback.INSTANCE,
                    AsyncAddArgs.newInstance(callback, ctx, System.currentTimeMillis(), byteBuf));
            return;
        }
```
What happened if asyncAddEntry fails from some reason? You lose the records `data` and you're notifying this to the callback